runtime/nvqir/CircuitSimulator.h

```diff
@@ -788,6 +804,44 @@ class CircuitSimulatorBase : public CircuitSimulator {
     return qubits;
   }
 
+  /// @brief Allocate `count` qubits in a specified state.
+  std::vector<std::size_t>
+  allocateQubits(std::size_t count,
+                 std::vector<std::complex<double>> &inState) override {
+    std::vector<std::size_t> qubits;
+    for (std::size_t i = 0; i < count; i++)
+      qubits.emplace_back(tracker.getNextIndex());
+
+    if (isInBatchMode()) {
+      // Store the current number of qubits requested
+      batchModeCurrentNumQubits += count;
+
+      // We have an allocated state, it has been set to |0>,
+      // we want to reuse it as is. If the state needs to grow, then
+      // we will ask the subtype to add more qubits.
+      if (qubits.back() < nQubitsAllocated)
+        count = 0;
+      else
+        count = qubits.back() + 1 - nQubitsAllocated;
+    }
+
+    cudaq::info("Allocating {} new qubits in a custom initial state.", count);
+
+    previousStateDimension = stateDimension;
+    nQubitsAllocated += count;
+    stateDimension = calculateStateDim(nQubitsAllocated);
+
+    // Tell the subtype to allocate more qubits
```

Comment:
This new `allocateQubits` is the same as the existing, but calls the new `addQubitsToState` overload that accepts the state data